### PR TITLE
Allow different types SourceName. 

### LINKED ISSUE #251: match error with SourceName
When ignoring a tile, it could be replaced with NODATA. However, the `SourceName` can be `StringName` then, and would cause issues in the FIleLayerProvider
`j-231219c859b44e60a252654db30210bb`

`Suppressed: scala.MatchError: (GridBoundsRasterRegion(MosaicRasterSource(List(org.openeo.geotrellis.layers.NoDataRasterSource(uint8, GridExtent(Extent(4561530.0, 3144110.0, 4678950.0, 3261370.0), CellSize(10.0,10.0), 11742x11726), EPSG:3035), org.openeo.geotrellis.layers.NoDataRasterSource(uint8, GridExtent(Extent(4561530.0, 3144110.0, 4678950.0, 3261370.0), CellSize(10.0,10.0), 11742x11726), EPSG:3035), org.openeo.geotrellis.layers.NoDataRasterSource(uint8, GridExtent(Extent(4561530.0, 3144110.0, 4678950.0, 3261370.0), CellSize(10.0,10.0), 11742x11726), EPSG:3035), org.openeo.geotrellis.layers.ValueOffsetRasterSource@5e9db88c, org.openeo.geotrellis.layers.ValueOffsetRasterSource@1899243b, org.openeo.geotrellis.layers.ValueOffsetRasterSource@62fb1f01, org.openeo.geotrellis.layers.NoDataRasterSource(uint8, GridExtent(Extent(4561530.0, 3144110.0, 4678950.0, 3261370.0), CellSize(10.0,10.0), 11742x11726), EPSG:3035), org.openeo.geotrellis.layers.ValueOffsetRasterSource@733bbf7f, org.openeo.geotrellis.layers.ValueOffsetRasterSource@65c60562, GDALRasterSource(/vsis3/EODATA/Sentinel-2/MSI/L2A/2018/10/28/S2B_MSIL2A_20181028T100119_N0206_R122_T33UVT_20181028T135751.SAFE/GRANULE/L2A_T33UVT_A008585_20181028T100117/IMG_DATA/R20m/T33UVT_20181028T100119_SCL_20m.jp2,GDALWarpOptions(-of VRT -r near -tap -tr 10.0 10.0 -t_srs +proj=laea +lat_0=52 +lon_0=10 +x_0=4321000 +y_0=3210000 +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs  -ovr AUTO -te 4561530.0 3144110.0 4678950.0 3261370.0 -te_srs +proj=laea +lat_0=52 +lon_0=10 +x_0=4321000 +y_0=3210000 +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs ))), EPSG:3035, GridExtent(Extent(4561530.0, 3144110.0, 4678950.0, 3261370.0), CellSize(10.0,10.0), 11742x11726), StringName(org.openeo.geotrellis.layers.NoDataRasterSource(uint8, GridExtent(Extent(4561530.0, 3144110.0, 4678950.0, 3261370.0), CellSize(10.0,10.0), 11742x11726), EPSG:3035))),GridBounds(9383,137,9894,648)),StringName(org.openeo.geotrellis.layers.NoDataRasterSource(uint8, GridExtent(Extent(4561530.0, 3144110.0, 4678950.0, 3261370.0), CellSize(10.0,10.0), 11742x11726), EPSG:3035))) (of class scala.Tuple2)`

### PR DESCRIPTION
**Pull request by EmileSonneveld**
Allow multiple types of SourceName
Type hierarchy:
```
OpenEoSourcePath -> SourcePath \
                    StringName --> SourceName
                     EmptyName /
```

### CODE REVIEW

File: openeo-geotrellis/src/main/scala/org/openeo/geotrellis/layers/FileLayerProvider.scala [hunk 1]
```diff
@@ -11,7 +11,7 @@ import geotrellis.raster.gdal.{GDALPath, GDALRasterSource, GDALWarpOptions}
...
 import geotrellis.raster.io.geotiff.OverviewStrategy
 import geotrellis.raster.rasterize.Rasterizer
-import geotrellis.raster.{CellSize, CellType, ConvertTargetCellType, CroppedTile, FloatConstantNoDataCellType, FloatConstantTile, GridBounds, GridExtent, MosaicRasterSource, MultibandTile, NoNoData, PaddedTile, Raster, RasterExtent, RasterMetadata, RasterRegion, RasterSource, ResampleMethod, ResampleTarget, ShortConstantNoDataCellType, SourceName, SourcePath, TargetAlignment, TargetCellType, TargetRegion, Tile, UByteUserDefinedNoDataCellType, UShortConstantNoDataCellType}
+import geotrellis.raster.{CellSize, CellType, ConvertTargetCellType, CroppedTile, FloatConstantNoDataCellType, FloatConstantTile, GridBounds, GridExtent, MosaicRasterSource, MultibandTile, NoNoData, PaddedTile, Raster, RasterExtent, RasterMetadata, RasterRegion, RasterSource, ResampleMethod, ResampleTarget, ShortConstantNoDataCellType, SourceName, EmptyName, SourcePath, StringName, TargetAlignment, TargetCellType, TargetRegion, Tile, UByteUserDefinedNoDataCellType, UShortConstantNoDataCellType}
```
Comment (bossie) [inline openeo-geotrellis/src/main/scala/org/openeo/geotrellis/layers/FileLayerProvider.scala]: `EmptyName` is unused.

File: openeo-geotrellis/src/main/scala/org/openeo/geotrellis/layers/FileLayerProvider.scala [hunk 2]
```diff
@@ -504,8 +504,8 @@ object FileLayerProvider {
       val allRegions = tuple._2.toSeq
 
       val tilesForRegion = allRegions
-        .flatMap { case (rasterRegion, sourcePath: SourcePath) =>
-          val result: Option[(MultibandTile, SourcePath)] = cloudFilterStrategy match {
+        .flatMap { case (rasterRegion, sourcePath: SourceName) =>
```
Comment (bossie) [inline openeo-geotrellis/src/main/scala/org/openeo/geotrellis/layers/FileLayerProvider.scala]: Replace `sourcePath` with `sourceName`.

File: openeo-geotrellis/src/main/scala/org/openeo/geotrellis/layers/FileLayerProvider.scala [hunk 3]
```diff
@@ -588,7 +588,19 @@ object FileLayerProvider {
...
+              case s: SourcePath => s.value
+              case s: StringName => s.value
+              case s => s.toString // ex: EmptyName
+            }
+            val rightValue = rightSourcePath match {
+              case s: SourcePath => s.value
+              case s: StringName => s.value
+              case s => s.toString // ex: EmptyName
+            }
+            leftValue < rightValue
```
Comment (bossie) [inline openeo-geotrellis/src/main/scala/org/openeo/geotrellis/layers/FileLayerProvider.scala]: Can probably be written shorter e.g.

```scala
            def sortable(sourceName: SourceName): String = sourceName match {
              case s: SourcePath => s.value
              case s: StringName => s.value
              case s => s.toString // ex: EmptyName
            }
            sortable(leftSourcePath) < sortable(rightSourcePath)
```

**Review by bossie — CHANGES_REQUESTED** [review]
[This commit](https://github.com/Open-EO/openeo-geotrellis-extensions/commit/f69f915f0eae51e53d764985ae8855e6b88e2514) can probably be reverted, right? Because you're handling all kinds of `SourceName` (`case s => s.toString`).

Please add a minimal test case. I wasn't able to find the process graph.

File: openeo-geotrellis/src/main/scala/org/openeo/geotrellis/layers/FileLayerProvider.scala (hunk 1 above)
Comment (EmileSonneveld) [reply openeo-geotrellis/src/main/scala/org/openeo/geotrellis/layers/FileLayerProvider.scala]: true, removed

File: openeo-geotrellis/src/main/scala/org/openeo/geotrellis/layers/FileLayerProvider.scala (hunk 3 above)
Comment (EmileSonneveld) [reply openeo-geotrellis/src/main/scala/org/openeo/geotrellis/layers/FileLayerProvider.scala]: implemented as `sortableSourceName`

File: openeo-geotrellis/src/main/scala/org/openeo/geotrellis/layers/FileLayerProvider.scala (hunk 2 above)
Comment (EmileSonneveld) [reply openeo-geotrellis/src/main/scala/org/openeo/geotrellis/layers/FileLayerProvider.scala]: done

Comment (EmileSonneveld) [reply openeo-geotrellis/src/main/scala/org/openeo/geotrellis/layers/FileLayerProvider.scala]: done

**Comment by EmileSonneveld** [comment]
I added the test process graph as a comment to the ticket

**Pull request merged**